Sources/ClientRuntime/Networking/Streaming/BufferedStream.swift

```diff
@@ -47,24 +51,42 @@ public class BufferedStream: Stream {
     /// - Parameter count: The maximum number of bytes to read.
     /// - Returns: Data read from the stream, or nil if the stream is closed and no data is available.
     public func read(upToCount count: Int) throws -> Data? {
-        lock.withLockingClosure {
-            let toRead = min(count, buffer?.count ?? 0)
-            let endPosition = position.advanced(by: toRead)
-            let chunk = buffer?[position..<endPosition]
+        try lock.withLockingThrowingClosure {
+            try _read(upToCount: count)
+        }
+    }
 
-            // remove the data we just read
-            buffer?.removeFirst(toRead)
+    private func _read(upToCount count: Int) throws -> Data? {
```

Comment:
Similar to another comment, we should use the lock in this function to prevent future bugs caused by human error.

Reply:
The intent is that `private func _read(upToCount:)` (as opposed to the `public` version, without the leading `_`) may be called when the lock is already held; this allows other methods that already hold the lock (such as `_serviceFirstReaderIfPossible()` to perform a read without deadlocking.

Reply:
`NSRecursiveLock` removes the risk of a deadlock in this case (which is why that type is being used for the lock instead of `NSLock`). It allows for the lock to be acquired multiple times by the same thread without causing a deadlock.